docs/proposals/20221007-ir-state.md

```diff
@@ -0,0 +1,81 @@
+# Refactor the IR to use state
+
+Author:
+- [kemingy](https://github.com/kemingy/)
+
+
+## Summary
+
+This proposal is designed to refactor the IR, as we discussed in [envd/#91](https://github.com/tensorchord/envd/issues/91#issuecomment-1239405359).
+
+## Motivation
+
+* Currently, we build the image with hard-coded order and parallelism. It's not flexible enough for some use cases.
+* The design of buildkit LLB state is suitable for method chaining implementation. Users can define the dependencies and parallelism easily.
+
+## Goals
+
+* Able to define the order and parallelism.
+* Provide low level APIs to build the image from scratch.
+
+## API
+
+```python
+def conda_env(root, version="3.10"):
+    conda = root.env("MAMBA_ROOT_PREFIX", "/opt/conda").run(
+        [
+            "curl micro.mamba.pm/install.sh | sh -s -- -y",
+            "/opt/conda/bin/micromamba create -n envd python={}".format(version),
+        ]
+    )
+    return conda
+
+
+def install_shell(root, sh="bash"):
+    if sh == "bash":
+        return root
+
+
+def install_vscode_ext(ext=()):
+    root = Source.merge(Source.http("openvsx/{}".format(x), filename=x) for x in ext)
+    return root
+
+
+def install_key():
+    root = (
+        Source.scratch()
```

Comment:
Could we remove the `Source`?

Use `scratch`, `merge` and `image`.

Reply:
fixed